internal/identity/gitlab.go

```diff
@@ -0,0 +1,118 @@
+package identity // import "github.com/pomerium/pomerium/internal/identity"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	oidc "github.com/pomerium/go-oidc"
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"github.com/pomerium/pomerium/internal/version"
+	"golang.org/x/oauth2"
+)
+
+const defaultGitLabProviderURL = "https://gitlab.com"
+const defaultGitLabGroupURL = "https://gitlab.com/api/v4/groups"
+const revokeURL = "https://gitlab.com/oauth/revoke"
+
+// GitLabProvider is an implementation of the OAuth Provider
+type GitLabProvider struct {
+	*Provider
+
+	RevokeURL string `json:"revocation_endpoint"`
+}
+
+// NewGitLabProvider returns a new GitLabProvider.
+// https://www.pomerium.io/docs/identity-providers.html#gitlab-active-directory
+func NewGitLabProvider(p *Provider) (*GitLabProvider, error) {
+	ctx := context.Background()
+
+	if p.ProviderURL == "" {
+		p.ProviderURL = defaultGitLabProviderURL
+	}
+
+	var err error
+	p.provider, err = oidc.NewProvider(ctx, p.ProviderURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(p.Scopes) == 0 {
+		p.Scopes = []string{oidc.ScopeOpenID, "api", "read_user", "profile", "email"}
+	}
+
+	p.verifier = p.provider.Verifier(&oidc.Config{ClientID: p.ClientID})
+	p.oauth = &oauth2.Config{
+		ClientID:     p.ClientID,
+		ClientSecret: p.ClientSecret,
+		Endpoint:     p.provider.Endpoint(),
+		RedirectURL:  p.RedirectURL.String(),
+		Scopes:       p.Scopes,
+	}
+	gp := &GitLabProvider{
+		Provider:  p,
+		RevokeURL: revokeURL,
+	}
+
+	if err := p.provider.Claims(&gp); err != nil {
+		return nil, err
+	}
+
+	p.UserGroupFn = gp.UserGroups
+
+	return gp, nil
+}
+
+// Revoke attempts to revoke session access via revocation endpoint
+// https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#revoking-a-personal-access-token
+func (p *GitLabProvider) Revoke(ctx context.Context, token *oauth2.Token) error {
+	params := url.Values{}
+	params.Add("access_token", token.AccessToken)
+
+	err := httputil.Client(ctx, http.MethodPost, p.RevokeURL, version.UserAgent(), nil, params, nil)
+	if err != nil && err != httputil.ErrTokenRevoked {
+		return err
+	}
+
+	return nil
+}
+
+// GetSignInURL returns the sign in URL with typical oauth parameters
+func (p *GitLabProvider) GetSignInURL(state string) string {
+	return p.oauth.AuthCodeURL(state, oauth2.AccessTypeOffline, oauth2.SetAuthURLParam("prompt", "select_account"))
+}
+
+// UserGroups returns a slice of groups for the user.
+//
+// By default, this request returns 20 results at a time because the API results are paginated.
+// https://docs.gitlab.com/ee/api/groups.html#list-groups
+func (p *GitLabProvider) UserGroups(ctx context.Context, s *sessions.State) ([]string, error) {
```

Comment:
I was unable to fetch groups.  Perhaps because I am not in any groups? However, I am getting a 200 response from the groups endpoint. 

<img width="541" alt="Screen Shot 2020-03-04 at 8 38 00 PM" src="https://user-images.githubusercontent.com/1544881/75948329-1030fb80-5e58-11ea-9947-3d4769ac8816.png">


When I add myself to a group, the json unforunately still does not parse

```
[{"id":7332779,"web_url":"https://gitlab.com/groups/RayStrong","name":"Ray Strong","path":"RayStrong","description":"","visibility":"private","share_with_group_lock":false,"require_two_factor_authentication":false,"two_factor_grace_period":48,"project_creation_level":"developer","auto_devops_enabled":null,"subgroup_creation_level":"maintainer","emails_disabled":null,"mentions_disabled":null,"lfs_enabled":true,"avatar_url":null,"request_access_enabled":true,"full_name":"Ray Strong","full_path":"RayStrong","parent_id":null,"ldap_cn":null,"ldap_access":null}]
```

```
8:51PM INF httputil: ErrorResponse error="Internal Server Error: oauth callback : error redeeming authenticate code: internal/identity: could not retrieve groups json: cannot unmarshal array into Go value of type struct { Groups []struct { ID string \"json:\\\"id\\\"\"; Name string \"json:\\\"name\\\"\"; Path string \"json:\\\"path\\\"\"; Description string \"json:\\\"description\\\"\"; Visibility string \"json:\\\"visibility\\\"\" } \"json:\\\"value\\\"\" }" ip=192.168.1.1 req_id=cc1aca1a-b339-d7d0-45fe-1d46d67594ba user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.122 Safari/537.36"
```

Reply:
I have worked on this and we should not have this issue again.